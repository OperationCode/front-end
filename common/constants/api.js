import { get, post, patch } from 'common/utils/api-utils';

/* GET REQUESTS */
export const getCodeSchoolsPromise = () => get('code_schools');
export const getMentorPromise = id => get(`mentors/${id}`);
export const getMentorsPromise = () => get('mentors');
export const getRequestsPromise = () => get('requests');
export const getScholarshipPromise = id => get(`scholarships/${id}`);
export const getScholarshipsPromise = () => get('scholarships');
export const getServicesPromise = () => get('services');
export const getTeamMembersPromise = () => get('team_members');

/* POST REQUESTS */
export const createUser = ({ email, password, firstName, lastName, zipcode }) =>
  post('users', {
    user: {
      email,
      password,
      first_name: firstName,
      last_name: lastName,
      zip: zipcode,
    },
  }).then(({ data }) => data);

export const loginUser = ({ email, password }) =>
  post('sessions', {
    user: {
      email,
      password,
    },
  }).then(({ data }) => data);

export const postMentorRequestPromise = ({ language, additionalDetails, mentor, service }) =>
  post('requests', {
    request: {
      details: additionalDetails,
      language,
      requested_mentor_id: mentor,
      service_id: service,
    },
  });

/* PATCH REQUESTS */
export const patchUpdateMentorRequestPromise = ({ request, status, mentor }) =>
  patch(`requests/${request}`, {
    request: {
      status,
      mentor,
    },
  });
