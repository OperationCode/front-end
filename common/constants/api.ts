import { patch, post } from 'common/utils/api-utils';
import type { RegistrationFormValues } from 'components/Forms/RegistrationForm/RegistrationForm';
import type { MilitaryStatusFormShape } from 'components/Forms/UpdateProfileForm/steps/MilitaryStatus';
import type { ProfessionalDetailsFormShape } from 'components/Forms/UpdateProfileForm/steps/ProfessionalDetails';

export const registerUser = (values: RegistrationFormValues) =>
  post('/api/registration/new', values);

export const updateUser = (values: ProfessionalDetailsFormShape | MilitaryStatusFormShape) =>
  patch('/api/registration/update', values);
