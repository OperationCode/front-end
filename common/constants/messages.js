export const validationErrorMessages = {
  required: 'Required',
  email: 'Must be a valid email',
  emailsMatch: 'Emails must match',
  emailExists: 'A user is already registered with this e-mail address',
  password:
    'Minimum password requirements: 8 characters, 1 Lowercase Letter, 1 Uppercase Letter, 1 Number',
  passwordsMatch: 'Passwords must match',
};

export const networkErrorMessages = {
  serverDown: 'Something is wrong on our end. Please try again later.',
};
