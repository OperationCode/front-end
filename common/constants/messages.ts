export const validationErrorMessages = {
  required: 'Required',
  email: 'Must be a valid email',
  emailsMatch: 'Emails must match',
  emailExists: 'A user is already registered with this e-mail address',
  codeOfConduct: 'You must agree to our Code of Conduct',
  slackGuidelines: 'You must agree to our Slack workspace guidelines',
};

export const networkErrorMessages = {
  serverDown: 'Something is wrong on our end. Please try again later.',
};
