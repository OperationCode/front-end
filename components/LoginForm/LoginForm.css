.LoginForm {
  width: 100%;
}

.row {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.topMargin {
  margin-top: 1rem;
}
