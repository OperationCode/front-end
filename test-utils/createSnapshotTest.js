/* eslint-env jest */
import renderer from 'react-test-renderer';

/**
 * Used to create fully-rendered snapshot test.
 * Useful for reusable component smoke tests.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
};
