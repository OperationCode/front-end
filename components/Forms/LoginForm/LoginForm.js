import React, { useState } from 'react';
import { func, shape, string } from 'prop-types';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';
import noop from 'lodash/noop';
import { getServerErrorMessage } from 'common/utils/api-utils';
import { validationErrorMessages } from 'common/constants/messages';
import { LOGIN_BUTTON, LOGIN_FORM } from 'common/constants/testIDs';
import Button from 'components/Buttons/Button/Button';
import Form from 'components/Form/Form';
import Input from 'components/Form/Input/Input';
import Alert from 'components/Alert/Alert';
import styles from './LoginForm.module.css';

/*
 * NOTE: We're repeating hardcode between the registration schema and passing an asterisk
 * to the label of required fields. This seems to an unfortunate negative aspect of an otherwise
 * awesome library. More importantly, it looks like the lib's author has plans to remedy the
 * situation. For now, our forms wont change much, so we should be okay.
 */
const loginSchema = Yup.object().shape({
  email: Yup.string()
    .required(validationErrorMessages.required)
    .email(validationErrorMessages.email),
  password: Yup.string().required(validationErrorMessages.required),
});

LoginForm.propTypes = {
  login: func.isRequired, // essentially onSubmit
  onSuccess: func.isRequired,
  initialValues: shape({
    email: string,
    password: string,
  }),
  redirectFunction: func,
  buttonTheme: string,
};

LoginForm.defaultProps = {
  initialValues: {
    email: '',
    password: '',
  },
  redirectFunction: noop,
  buttonTheme: 'secondary',
};

function LoginForm({ initialValues, login, onSuccess, redirectFunction, buttonTheme }) {
  const [errorMessage, setErrorMessage] = useState('');

  const handleSubmit = async (values, actions) => {
    try {
      const { token } = await login(values);

      await onSuccess({ token });
      actions.setSubmitting(false);
      actions.resetForm();
      redirectFunction();
    } catch (error) {
      actions.setSubmitting(false);

      setErrorMessage(getServerErrorMessage(error));
    }
  };

  return (
    <Formik initialValues={initialValues} onSubmit={handleSubmit} validationSchema={loginSchema}>
      {({ isSubmitting }) => (
        <Form className={styles.LoginForm} data-testid={LOGIN_FORM}>
          <div className={styles.row}>
            <Field
              type="email"
              name="email"
              label="Email*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="username email"
            />

            <Field
              type="password"
              name="password"
              label="Password*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="new-password"
            />

            {errorMessage && <Alert type="error">{errorMessage}</Alert>}

            <Button
              className={styles.topMargin}
              type="submit"
              theme={buttonTheme}
              disabled={isSubmitting}
              data-testid={LOGIN_BUTTON}
            >
              Login
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
}

export default LoginForm;
