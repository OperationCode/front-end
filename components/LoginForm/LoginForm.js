import React, { useState } from 'react';
import { func, shape, string } from 'prop-types';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';
import { getServerErrorMessage } from 'common/utils/api-utils';
import { validationErrorMessages } from 'common/constants/messages';
import Button from 'components/Buttons/Button/Button';
import Form from 'components/Form/Form';
import Input from 'components/Form/Input/Input';
import Alert from 'components/Alert/Alert';
import styles from './LoginForm.module.css';

/*
 * NOTE: We're repeating hardcode between the registration schema and passing an asterisk
 * to the label of required fields. This seems to an unfortunate negative aspect of an otherwise
 * awesome library. More importantly, it looks like the lib's author has plans to remedy the
 * situation. For now, our forms wont change much, so we should be okay.
 */
const loginSchema = Yup.object().shape({
  email: Yup.string()
    .required(validationErrorMessages.required)
    .email(validationErrorMessages.email),
  password: Yup.string().required(validationErrorMessages.required),
});

LoginForm.propTypes = {
  login: func.isRequired, // essentially onSubmit
  onSuccess: func.isRequired,
  initialValues: shape({
    email: string,
    password: string,
  }),
};

LoginForm.defaultProps = {
  initialValues: {
    email: '',
    password: '',
  },
};

function LoginForm({ initialValues, login, onSuccess }) {
  const [errorMessage, setErrorMessage] = useState('');

  const handleSubmit = async (values, actions) => {
    try {
      const { token } = await login(values);

      await onSuccess({ token });
      actions.setSubmitting(false);
      actions.resetForm();
    } catch (error) {
      actions.setSubmitting(false);

      setErrorMessage(getServerErrorMessage(error));
    }
  };

  return (
    <Formik initialValues={initialValues} onSubmit={handleSubmit} validationSchema={loginSchema}>
      {({ isSubmitting }) => (
        <Form className={styles.LoginForm}>
          <div className={styles.row}>
            <Field
              type="email"
              name="email"
              label="Email*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="username email"
            />

            <Field
              type="password"
              name="password"
              label="Password*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="new-password"
            />

            {errorMessage && <Alert type="error">{errorMessage}</Alert>}

            <Button
              className={styles.topMargin}
              type="submit"
              theme="secondary"
              disabled={isSubmitting}
            >
              Submit
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
}

export default LoginForm;
