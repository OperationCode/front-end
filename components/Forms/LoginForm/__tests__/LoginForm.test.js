import faker from 'faker';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { loginUser } from 'common/constants/api';
import { networkErrorMessages, validationErrorMessages } from 'common/constants/messages';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import OperationCodeAPIMock from 'test-utils/mocks/apiMock';
import mockUser from 'test-utils/mockGenerators/mockUser';
import LoginForm from '../LoginForm';

describe('LoginForm', () => {
  afterEach(() => {
    OperationCodeAPIMock.reset();
  });

  it('should render with required props', () => {
    createSnapshotTest(<LoginForm login={jest.fn()} onSuccess={jest.fn()} />);
  });

  it('should display required error message when blurring past email input', async () => {
    render(<LoginForm login={jest.fn()} onSuccess={jest.fn()} />);

    fireEvent.blur(screen.getByLabelText(/Email/));

    expect(await screen.findByText(validationErrorMessages.required)).toBeInTheDocument();
  });

  it('should show error when providing non-email to email input', async () => {
    render(<LoginForm login={jest.fn()} onSuccess={jest.fn()} />);

    fireEvent.change(screen.getByLabelText(/Email/), { target: { value: 'email' } });
    fireEvent.blur(screen.getByLabelText(/Email/));

    expect(await screen.findByText(validationErrorMessages.email)).toBeInTheDocument();
  });

  it('should show "password required" message when blurring past input', async () => {
    render(<LoginForm login={jest.fn()} onSuccess={jest.fn()} />);

    fireEvent.blur(screen.getByLabelText(/Password/));

    expect(await screen.findByText(validationErrorMessages.required)).toBeInTheDocument();
  });

  it('should submit with valid data in form', async () => {
    const user = mockUser();

    const initialValues = {
      email: user.email,
      password: user.password,
    };

    OperationCodeAPIMock.onPost('auth/login/', initialValues).reply(200, {
      user: {
        ...user,
        slackName: faker.internet.userName(),
        mentor: false,
      },
      token: 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9',
    });

    const successSpy = jest.fn();
    render(<LoginForm onSuccess={successSpy} login={loginUser} initialValues={initialValues} />);

    fireEvent.click(screen.getByText('Login'));

    await waitFor(() => expect(successSpy).toHaveBeenCalledTimes(1));
    await waitFor(() => expect(OperationCodeAPIMock.history.post.length).toBeGreaterThan(0));
  });

  it('should NOT submit with invalid data in form', async () => {
    const successSpy = jest.fn();

    render(
      <LoginForm
        onSuccess={successSpy}
        login={jest.fn()}
        initialValues={{ email: 'email@email', password: '1' }}
      />,
    );

    fireEvent.click(screen.getByText('Login'));

    await waitFor(() => expect(successSpy).not.toHaveBeenCalled());
    await waitFor(() => expect(OperationCodeAPIMock.history.post.length).not.toBeGreaterThan(0));
  });

  it('should show error when trying to login with incorrect email or password', async () => {
    const invalidError = 'Invalid Email or password.';

    const initialValues = {
      email: 'testing123@gmail.com',
      password: 'Testing123',
    };

    OperationCodeAPIMock.onPost('auth/login/', initialValues).reply(401, { error: invalidError });

    const successSpy = jest.fn();

    render(<LoginForm login={loginUser} onSuccess={successSpy} initialValues={initialValues} />);

    fireEvent.click(screen.getByText('Login'));

    expect(await screen.findByText(invalidError)).toBeInTheDocument();
    await waitFor(() => expect(successSpy).not.toHaveBeenCalled());
  });

  it('should show a helpful error if the server is down', async () => {
    const user = mockUser();

    const initialValues = {
      email: user.email,
      password: user.password,
    };

    OperationCodeAPIMock.onPost('auth/login/', initialValues).reply(503);

    const successSpy = jest.fn();
    render(<LoginForm onSuccess={successSpy} login={loginUser} initialValues={initialValues} />);

    fireEvent.click(screen.getByText('Login'));

    await waitFor(() =>
      expect(screen.getByText(networkErrorMessages.serverDown)).toBeInTheDocument(),
    );
    await waitFor(() => expect(successSpy).not.toHaveBeenCalled());
    await waitFor(() => expect(OperationCodeAPIMock.history.post.length).toBeGreaterThan(0));
  });

  it('should reset form and set form as "not submitting" after successful login', async () => {
    const user = mockUser();

    const initialValues = {
      email: user.email,
      password: user.password,
    };

    OperationCodeAPIMock.onPost('auth/login/', initialValues).reply(200, {
      user: {
        ...user,
        slackName: faker.internet.userName(),
        mentor: false,
      },
      token: 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9',
    });

    const successSpy = jest.fn(() => Promise.resolve(true));
    render(<LoginForm onSuccess={successSpy} login={loginUser} initialValues={initialValues} />);

    expect(successSpy).toHaveBeenCalledTimes(0);

    fireEvent.click(screen.getByText('Login'));

    await waitFor(() => expect(successSpy).toHaveBeenCalledTimes(1));
    await waitFor(() => expect(screen.getByText('Login')).not.toBeDisabled());

    await waitFor(() => expect(screen.getByLabelText('Email*')).toHaveValue(''));
    await waitFor(() => expect(screen.getByLabelText('Password*')).toHaveValue(''));
  });
});
