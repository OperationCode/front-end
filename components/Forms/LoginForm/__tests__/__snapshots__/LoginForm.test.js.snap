// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should render with required props 1`] = `
<form
  className="LoginForm"
  data-testid="LOGIN_FORM"
  noValidate={true}
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="field"
      data-testid="INPUT"
    >
      <label
        className="Label"
        data-testid="LABEL"
        htmlFor="email"
      >
        Email*
      </label>
      <div
        className="inputFeedbackGrouping"
        data-testid="INPUT_FEEDBACK_GROUPING"
      >
        <input
          autoComplete="username email"
          className="Input"
          disabled={false}
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
      data-testid="INPUT"
    >
      <label
        className="Label"
        data-testid="LABEL"
        htmlFor="password"
      >
        Password*
      </label>
      <div
        className="inputFeedbackGrouping"
        data-testid="INPUT_FEEDBACK_GROUPING"
      >
        <input
          autoComplete="new-password"
          className="Input"
          disabled={false}
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
    
    <button
      className="Button topMargin secondary"
      data-testid="LOGIN_BUTTON"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="submit"
    >
      Login
    </button>
  </div>
</form>
`;
